### PR TITLE
`getAttestationRewards` API endpoint

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implements https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Rewards/getAttestationsRewards

The current implementation works for Altair and later forks and does not support inclusion delay.

**Which issues(s) does this PR fix?**

Part of #11952 

**Other notes for review**

Flow:

1. If epoch is before Altar, return 404.
2. If epoch is current or previous epoch, return 404.
3. Get slot at slot 31 of of `epoch+1`.
4. Initialize precompute balances and validators, process epoch participation.
5. Filter validators based on request body (use all validators when no body is present).
6. Create 32 hypothetical validators with effective balance from 1 to 32, setting up each of them in such a way that they will earn the maximum possible reward, and call `AttestationsDelta` with these validators and the precompute balance.
7. Call `AttestationsDelta` with the filtered validators and the precompute balance.
8. Populate and return rewards.

~~As you can see below, for some reason our rewards are smaller than Lightouse's. I did not include the ideal reward for `0` balance because it's useless.~~

### CODE REVIEW

File: beacon-chain/rpc/eth/rewards/handlers.go [hunk 1]
```diff
@@ -166,22 +178,318 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
...
+		filteredVals = make([]*precompute.Validator, len(valIndices))
+		for i, valIx := range valIndices {
+			filteredVals[i] = allVals[valIx]
+		}
+	}
+
+	idealRewards := make([]AttReward, 32)
+	idealVals := make([]*precompute.Validator, 32)
+	increment := params.BeaconConfig().EffectiveBalanceIncrement
+	for i := 1; i < 33; i++ {
```
Comment (potuz) [inline beacon-chain/rpc/eth/rewards/handlers.go:314]: Better to use `MaxEffectiveBalance` here and in the previous line to obtain 32 and 32+1

File: beacon-chain/rpc/eth/rewards/handlers.go [hunk 2]
```diff
@@ -166,22 +178,318 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
...
+		err = attestationDelta(r, bal, vals[i], baseRewardMultiplier, inactivityDenominator, leak)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func attestationDelta(
```
Comment (potuz) [inline beacon-chain/rpc/eth/rewards/handlers.go:400]: Please rename to `singleAttDelta` or something that is a bit more distinguishable from the plural version

File: beacon-chain/rpc/eth/rewards/handlers.go [hunk 3]
```diff
@@ -166,22 +178,318 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
...
+	baseReward := (effectiveBalance / increment) * baseRewardMultiplier
+	activeIncrement := bal.ActiveCurrentEpoch / increment
+
+	weightDenominator := cfg.WeightDenominator
+	srcWeight := cfg.TimelySourceWeight
+	tgtWeight := cfg.TimelyTargetWeight
+	headWeight := cfg.TimelyHeadWeight
+	// Process source reward / penalty
+	if val.IsPrevEpochSourceAttester && !val.IsSlashed {
+		if inactivityLeak {
```
Comment (potuz) [inline beacon-chain/rpc/eth/rewards/handlers.go:425]: The different logic here and in the corresponding `attestationDelta` function in the `precompute` package makes this hard to review and is bug prone.

File: beacon-chain/rpc/eth/rewards/handlers.go [hunk 4]
```diff
@@ -166,22 +178,318 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
...
+		}
+		network.WriteError(w, errJson)
+		return
+	}
+	currentEpoch := uint64(slots.ToEpoch(s.TimeFetcher.CurrentSlot()))
+	if epoch >= currentEpoch-1 {
+		errJson := &network.DefaultErrorJson{
+			Code:    http.StatusBadRequest,
+			Message: "Attestation rewards are available after two epoch transitions to ensure all attestation have a chance of inclusion",
+		}
```
Comment (kevink1103) [inline beacon-chain/rpc/eth/rewards/handlers.go:206]:
```suggestion
	if primitives.Epoch(epoch) < params.BeaconConfig().AltairForkEpoch {
		errJson := &network.DefaultErrorJson{
			Message: "Attestation rewards are not supported for Phase 0",
			Code:    http.StatusNotFound,
		}
		network.WriteError(w, errJson)
		return
	}
	currentEpoch := uint64(slots.ToEpoch(s.TimeFetcher.CurrentSlot()))
	if epoch >= currentEpoch-1 {
		errJson := &network.DefaultErrorJson{
			Code:    http.StatusNotFound,
			Message: "Attestation rewards are available after two epoch transitions to ensure all attestation have a chance of inclusion",
		}
```

> If epoch is before Altar, return 404.
> If epoch is current or previous epoch, return 404.

Looks like the return status code does not satisfy the spec in your description.

File: beacon-chain/rpc/eth/rewards/handlers.go [hunk 5]
```diff
@@ -166,22 +178,318 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
...
+	epochStart, err := slots.EpochStart(primitives.Epoch(epoch + 2))
+	if err != nil {
+		errJson := &network.DefaultErrorJson{
+			Message: "Could not get epoch's starting slot: " + err.Error(),
+			Code:    http.StatusInternalServerError,
+		}
+		network.WriteError(w, errJson)
+		return
+	}
+	st, err := s.Stater.StateBySlot(r.Context(), epochStart)
```
[the remark this answers is not in the document]
Comment (rkapka) [reply beacon-chain/rpc/eth/rewards/handlers.go:220]: it uses the state replayer internally, which does process slots

File: beacon-chain/rpc/eth/rewards/handlers.go (hunk 1 above)
Comment (rkapka) [reply beacon-chain/rpc/eth/rewards/handlers.go:314]: `MaxEffectiveBalance` is equal `32 * 1e9`

File: beacon-chain/rpc/eth/rewards/handlers.go (hunk 4 above)
Comment (rkapka) [reply beacon-chain/rpc/eth/rewards/handlers.go:206]: Thanks for noticing. I made a mistake in the description.

File: beacon-chain/rpc/eth/rewards/handlers.go (hunk 2 above)
Comment (kevink1103) [reply beacon-chain/rpc/eth/rewards/handlers.go:400]: https://github.com/prysmaticlabs/prysm/blob/180f310d5393e50c4628786d877bc93b337dde3f/beacon-chain/core/altair/epoch_precompute.go#L256-L293

Seems like the function names are same as here.

**Comment by kevink1103** [comment]
I've used your implementation on my node to see the return values and they are actually identical to what's shown on beaconcha.in (.
For instance,
`eth/v1/beacon/rewards/attestations/208350` with `["563065"]` got me this:
```json
"validator_index": "563065",
"head": "3137",
"target": "5898",
"source": "3177",
"inclusion_delay": ""
```
and on [beaconcha.in](https://beaconcha.in/validator/563065),
image
image
As you can see, the results from prysm and beaconcha.in are identical with 1 epoch difference (208350 on prysm, 208351 on beaconcha.in)

**Comment by rkapka** [comment]
Hey @kevink1103 , thanks a lot for looking into this. Yes, the issue is the current code fetches the incorrect state. It should get the state at slot 31 of epoch N+1 instead of at slot 0 of epoch N+2. I have the fix locally and will push my changes later today. They will include splitting the code into smaller functions and unit tests.

File: beacon-chain/rpc/eth/rewards/structs.go [hunk 6]
```diff
@@ -14,3 +14,29 @@ type BlockRewards struct {
...
+
+type AttestationRewardsResponse struct {
+	Data                AttestationRewards `json:"data"`
+	ExecutionOptimistic bool               `json:"execution_optimistic"`
+	Finalized           bool               `json:"finalized"`
+}
+
+type AttestationRewards struct {
+	IdealRewards []IdealAttestationReward
+	TotalRewards []TotalAttestationReward
```
Comment (kevink1103) [inline beacon-chain/rpc/eth/rewards/structs.go:26]:
```suggestion
	IdealRewards []IdealAttestationReward `json:"ideal_rewards"`
	TotalRewards []TotalAttestationReward `json:"total_rewards"`
```

I think this should be in snake_case.

File: beacon-chain/rpc/eth/rewards/handlers.go [hunk 7]
```diff
@@ -166,22 +175,285 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
...
+			Code:    http.StatusNotFound,
+		}
+		network.WriteError(w, errJson)
+		return nil, false
+	}
+	currentEpoch := uint64(slots.ToEpoch(s.TimeFetcher.CurrentSlot()))
+	if requestedEpoch+1 >= currentEpoch {
+		errJson := &network.DefaultErrorJson{
+			Code:    http.StatusNotFound,
+			Message: "Attestation rewards are available after two epoch transitions to ensure all attestations have a chance of inclusion",
```
Comment (kevink1103) [inline beacon-chain/rpc/eth/rewards/handlers.go:252]: Not sure if it's a must but should message be same with the [API spec](https://ethereum.github.io/beacon-APIs/#/Rewards/getAttestationsRewards)?
ex) "Epoch not known or required data not available" for 404

**Review by kevink1103 — COMMENTED** [review]
thank you for the work! just deployed your code on my node and everything works great!

File: beacon-chain/rpc/eth/rewards/handlers.go [hunk 8]
```diff
@@ -166,22 +175,285 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
...
+			}
+			valIndices[i] = primitives.ValidatorIndex(index)
+		}
+	}
+	if len(valIndices) == 0 {
+		valIndices = make([]primitives.ValidatorIndex, len(allVals))
+		for i := 0; i < len(allVals); i++ {
+			valIndices[i] = primitives.ValidatorIndex(i)
+		}
+	}
```
Comment (potuz) [inline beacon-chain/rpc/eth/rewards/handlers.go:352]: isn't this being the default a little bit of a DoS vector?

File: beacon-chain/rpc/eth/rewards/handlers.go (hunk 1 above)
Comment (potuz) [reply beacon-chain/rpc/eth/rewards/handlers.go:368]: I still don't like these unnamed constants

File: beacon-chain/rpc/eth/rewards/handlers.go [hunk 9]
```diff
@@ -166,22 +175,285 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
...
+	} else {
+		filteredVals := make([]*precompute.Validator, len(valIndices))
+		for i, valIx := range valIndices {
+			filteredVals[i] = allVals[valIx]
+		}
+		return bal, filteredVals, valIndices, true
+	}
+}
+
+func idealAttRewards(w http.ResponseWriter, st state.BeaconState, bal *precompute.Balance) ([]IdealAttestationReward, bool) {
```
Comment (potuz) [inline beacon-chain/rpc/eth/rewards/handlers.go:364]: A description of what this function does would be helpful

**Review by potuz — APPROVED** [review]
I vouched this time only the core logic and it looks OK to me. Left a couple of comments. I do have one extra: it seems to me that this endpoint is overly verbose for most validators, could we not filter the ideal rewards for the effective balances of those validators that were requested? Most of the time this would be 32ETH.

At any rate, I think you want to remove the balances 1--16 since such validators would be ejected anyway

File: beacon-chain/rpc/eth/rewards/handlers.go (hunk 8 above)
Comment (rkapka) [reply beacon-chain/rpc/eth/rewards/handlers.go:352]: I don't think DoS-ing is that much of a concern in Beacon APIs, but you may want to raise an issue in the spec repo: https://github.com/ethereum/beacon-APIs

**Review by potuz — APPROVED** [review]

**Pull request merged**